### ISSUE TITLE
Some fire/smoke effect broken since alpha 30 (vertexpack merge)

### ISSUE THREAD

**Issue by illwieckz** [Unvanquished/Unvanquished#780]
Hi, while porting the ut-metro map by newin, I've discovered that some fire and smoke effect that were still working in Unvanquished 0.29.0 were not working since Unvanquished 0.30.0. You know, the alpha 30 release is the one where the `vertexpack` branch were merged.

This is how fire and smoke looks like on Daemon 0.29.0 and before:
![fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.27.0/unvanquished_2015-07-27_012637_000.jpg)
![fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.27.0/unvanquished_2015-07-27_013048_000.jpg)
![fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.27.0/unvanquished_2015-07-27_013536_000.jpg)
![fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.27.0/unvanquished_2015-07-27_014627_000.jpg)

This is how fire and smoke (does not) look like on Daemon 0.30.0 and after:
![fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.41.0/unvanquished_2015-07-27_014931_000.jpg)
![fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.41.0/unvanquished_2015-07-27_015114_000.jpg)
![fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.41.0/unvanquished_2015-07-27_015212_000.jpg)
![fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.41.0/unvanquished_2015-07-27_015135_000.jpg)

This is not the only issue in these screenshots (you can see a blend issues on all screenshots, see the dark zones), but this is not on the scope of this ticket.

The fire is completely not rendered, the smoke is sometime, a little, rendered, but not like it was before.

I can send a preview pk3 of this map if needed.

**Comment by illwieckz** [comment]
The map can be found here: https://github.com/interstellar-oasis/map-terminus :wink:

**Comment by illwieckz** [comment]
It seems related to the `clamp` keyword.

This is an example of fire shader used in this map:

```
textures/terminus/mon_fire
{
	qer_editorimage textures/terminus/fire
	q3map_cloneshader textures/terminus/zzmon_fire_1
	surfaceparm nonsolid
	surfaceparm trans
	surfaceparm nomarks
	surfaceparm nodlight
	deformvertexes move 0 0 35 sawtooth 0 1 0 0.784
	deformvertexes move -0.755352 1.22297 0 inversesawtooth -0.695684 1 0 0.784
	deformvertexes autosprite
	cull disable
	{
		clampmap textures/terminus/fire
		blendfunc GL_SRC_ALPHA GL_ONE
		rgbgen wave sawtooth 1 -1 1 0.784
		alphagen wave sawtooth 0 6 0 0.784
		tcMod rotate -33
		tcMod stretch sawtooth 1 -0.7 0 0.784
	}
}
```

As you see, this shader (and all others) use the `clampmap` keyword:

```
		clampmap textures/terminus/fire
```

Perhaps it's not an xreal shader keyword since the [xreal shader manual does not mention it](http://tremap.xtr3m.net/__Games/Xreal/Manual_Shader_1/ShaderManual.htm) (but it worked on Unvanquished ≤ 0.29), instead there is a [`clamp` keyword](http://tremap.xtr3m.net/__Games/Xreal/Manual_Shader_1/ShaderManual.htm#clamp), that is used with a `map` keyword ([see usage example](http://tremap.xtr3m.net/__Games/Xreal/Manual_Shader_1/ShaderManual.htm#stageheatHazeMap)).

So, if I rewrite this line like that:

```
		map textures/terminus/fire
		clamp
```

I get the same result (no fire effect).

But if I comment out the `clamp` keyword:

```
		map textures/terminus/fire
		// clamp
```

I get a broken fire effect, so it seems other shader lines are working:

![fire](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.41.0/unvanquished_2015-08-01_025325_000.jpg)

If you want to try it yourself, you can tweak `textures/terminus_fx.shader` [from this package](http://gg.illwieckz.net/dl/unvanquished/pkg/map-terminus_b2~1.pk3).

Note: I don’t know if the shader in the picture is the one I show you the code above, I applied the change to all shaders to test them.
Note: The black texture issue is another unrelated issue (see #782).

**Comment by illwieckz** [comment]
Another screenshot without `clamp` keyword:

![fire](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.41.0/unvanquished_2015-08-01_025950_000.jpg)

**Comment by illwieckz** [comment]
For information, the `clampmap` keyword is still supported by the current code base:

https://github.com/Unvanquished/Unvanquished/blob/38f0b762c14cb2df86fd5baa1d7e9624d75d579d/daemon/src/engine/renderer/tr_shader.cpp#L1605

**Comment by illwieckz** [comment]
@gimhael, this regression was introduced by commit 0066c5fc60660bba75ca64ece539ca70f85b9e4a

```
Date:   Sat Aug 2 10:31:41 2014 +0200
Use the same vertex layout in tess and VBOs to simplify upload.
```

**Comment by illwieckz** [comment]
Probably not related at all to the clampmap keyword, it's probably just a side effect.

**Comment by illwieckz** [comment]
This bug was fixed in 0.49.0 release.

![[fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035618_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035618_000.jpg)
![[fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035626_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035626_000.jpg)
![[fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035629_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035629_000.jpg)
![[fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035635_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035635_000.jpg)
![[fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035741_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035741_000.jpg)
![[fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035801_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035801_000.jpg)
![[fire and smoke](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035808_000.jpg)](http://dl.illwieckz.net/b/unvanquished/bugs/fire-ut-metro/unvanquished_0.50.0/unvanquished_2016-04-11_035808_000.jpg)
